### PR TITLE
[Merged by Bors] - feat: the Faa di Bruno formula

### PR DESCRIPTION
**Pull request by sgouezel**
---

- [x] depends on: #17169

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean [hunk 1]
```diff
@@ -683,4 +685,298 @@ def extendEquiv (n : ℕ) :
...
+    c.applyOrderedFinpartition (update p m q) v
+      = update (c.applyOrderedFinpartition p v) m (q (v ∘ c.emb m)) := by
+  ext d
+  by_cases h : d = m
+  · rw [h]
+    simp [applyOrderedFinpartition]
+  · simp [h, applyOrderedFinpartition]
+
+/-- Given a an ordered finite partition `c` of `n`, a continuous multilinear map `f` in `c.length`
+variable, and for each `m` a continuous multilinear map `p m` in `c.partSize m` variables,
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean:740]:
```suggestion
variables, and for each `m` a continuous multilinear map `p m` in `c.partSize m` variables,
```

File: Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean [hunk 2]
```diff
@@ -683,4 +685,298 @@ def extendEquiv (n : ℕ) :
...
+    (q.compAlongOrderedFinpartition p c) v =
+      q c.length (c.applyOrderedFinpartition (fun m ↦ (p (c.partSize m))) v) :=
+  rfl
+
+/-- Taylor formal composition of two formal multilinear series. The `n`-th coefficient in the
+composition is defined to be the sum of `q.compAlongOrderedFinpartition p c` over all
+ordered partitions of `n`. In other words, this term (as a multilinear function applied to
+`v₀, ..., vₙ₋₁`) is
+`∑'_{k} ∑'_{I₀ ⊔ ... ⊔ Iₖ₋₁ = {0, ..., n-1}} qₖ (p_{i₀} (...), ..., p_{iₖ₋₁} (...))`, where
+`i_m` is the size of `I_m` and one puts all variables of `I_m` as arguments to `p_{iₘ}`, in
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean:834]: Do you want to use unicode subscripts here?

File: Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean [hunk 3]
```diff
@@ -683,4 +685,298 @@ def extendEquiv (n : ℕ) :
...
+  apply B.analyticOnNhd_uncurry_of_multilinear.comp_analyticOn ?_ (mapsTo_univ _ _)
+  apply AnalyticOn.prod
+  · exact (hq c.length).comp hf h
+  · exact AnalyticOn.pi (fun i ↦ hp _)
+
+open OrderedFinpartition
+
+/-- Composing two formal multilinear series `q` and `p` along an ordered partition extended by a
+new atom to the left corresponds to applying `p 1` on the first coordinates, and the initial
+ordered partition on the other corodinates.
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean:868]:
```suggestion
ordered partition on the other coordinates.
```

File: Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean [hunk 4]
```diff
@@ -683,4 +685,298 @@ def extendEquiv (n : ℕ) :
...
+This is one of the terms that appears when differentiating in the Faa di Bruno
+formula, going from step `m` to step `m + 1`. -/
+lemma faaDiBruno_aux1 {m : ℕ} (q : FormalMultilinearSeries 𝕜 F G)
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean:871]: Maybe make those private

File: Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean [hunk 5]
```diff
@@ -683,4 +685,298 @@ def extendEquiv (n : ℕ) :
...
+    apply FormalMultilinearSeries.congr _ (by simp [hij])
+    simp
+
+/-- *Faa di Bruno* formula: If two functions `g` and `f` have Taylor series up to `n` given by
+`q` and `p`, then `g ∘ f` also has a Taylor series, given by `q.taylorComp p`. -/
+theorem HasFTaylorSeriesUpToOn.comp {n : ℕ∞} {g : F → G} {f : E → F}
+    (hg : HasFTaylorSeriesUpToOn n g q t) (hf : HasFTaylorSeriesUpToOn n f p s) (h : MapsTo f s t) :
+    HasFTaylorSeriesUpToOn n (g ∘ f) (fun x ↦ (q (f x)).taylorComp (p x)) s := by
+  /- One has to check that the `m+1`-th term is the derivative of the `m`-th term. The `m`-th term
+  is a sum, that one can differentiate terms by terms. Each term is a linear map into continuous
```
Comment (jcommelin) [inline Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean:924]:
```suggestion
  is a sum, that one can differentiate term by term. Each term is a linear map into continuous
```

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**